Dev authorization fakes

DarthPedro opened the pull request:
## Pull request description
For components that use the authentication and authorization, there is a lot of boilerplate code required to write to create a test. This code can be complex and needs to be copied into user's test projects each time.

Provide a set of "fake" authorization services that allow the user to choose and change the authentication/authorization state and user identification for a test. (closes #146)

### PR meta checklist
- [x] Pull request is targeting at `DEV` branch.
- [x] Pull request is linked to all related issues, if any.
- [x] I have read the _CONTRIBUTING.md_ document.

### Content checklist
- [x] My code follows the code style of this project and AspNetCore coding guidelines.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] I have updated the appropriate sub section in the _CHANGELOG.md_.
- [x] I have added, updated or removed tests to according to my changes.
  - [x] All tests passed.

### Todo checklist
- [x] Does this cover all auth scenarios, e.g. both declarative in razor syntax and in c# code?
- [x] Are there other types of claims that should be supported besides just basic/username? Is it relevant for testing? [DarthPedro] other claims can be really mocked based on what's required for a component or test. It shouldn't need fakes for that at this point. Blazor components like AuthorizeView and AuthorizeRouteView don't make use of claims.
- [x] Test that we can correctly switch between authenticated and unauthenticated states.
- [x] Create unit tests that covers the FakeAuthenticationStateProvider and FakeAuthorizationService. This can most likely be a regular unit test, maybe with a single component test in the mix to verify that e.g. AuthorizeView is compatible.
- [x] Create a document page that describe how to use the FakeAuthenticationStateProvider
- [x] Create a stub for FakeAuthenticationStateProvider (#115) that is registered by default and provides the user with a good error message that tells how to register FakeAuthenticationStateProvider and link to docs. See example from JSMock at:
  - [x] PlaceholderAuthenticationStateProvider.cs and PlaceholderAuthorizationService.cs.
  - [x] TestServiceProviderExtensions.cs register the stubs in the AddDefaultTestContextServices.
- [x] Code should be placed in src/bunit.web/TestDoubles/Authorization or something similar.
- [x] Test code placement should mirror source code in src/bunit.web.test/TestDoubles/Authorization.
- [x] Create extension methods for TestServiceProvider which supports registering with or without a username, e.g. Services.AddAuthorization() and Service.AddAuthorization("UserName", isAuthorized = true);

egil commented on src/bunit.web/TestDoubles/Authorization/FakeAuthenticationStateProvider.cs:
```diff
@@ -0,0 +1,81 @@
+namespace Bunit.TestDoubles.Authorization
+{
+	/// <summary>
+	/// Test class that provides an implementation of AuthenticationStateProvider that allows
```
```suggestion
	/// Represents a fake implementation of AuthenticationStateProvider for testing purposes that allows
```

egil commented on src/bunit.web/TestDoubles/Authorization/FakeAuthenticationStateProvider.cs:
```diff
@@ -0,0 +1,81 @@
+		/// <param name="roles">Roles that this user principal has.</param>
+		public FakeAuthenticationStateProvider(string username, IList<string>? roles = null)
+		{
+			this.CurrentAuthStateTask = FakeAuthenticationStateProvider.CreateAuthenticationState(username, roles);
```
I like to avoid `this.` unless the compiler complaints

DarthPedro replied:
Sorry, that's a coding style that I am used to. I will probably have to change it in many places. I'll take a look at that.

egil replied:
No worries. Think you can use a refactoring in VS that will do it for you. A quick search/replace at least.

DarthPedro replied:
Updated to remove this prefix.

egil commented on src/bunit.web/TestDoubles/Authorization/FakeIdentity.cs:
```diff
@@ -0,0 +1,25 @@
+		/// <summary>
+		/// Gets the test authentication type.
+		/// </summary>
+		public string AuthenticationType => "Test";
```
Should this be setable by the user?

DarthPedro replied:
The authentication type has not been important to Blazor auth process, but I can make it settable. Would you like me to make that change? Most users will just use the AddAuthorization extension methods and never access these lower level types.

egil replied:
Does it give us any problems making it settable with a default value?

I like to build APIs that are as intuitive as possible, with as much flexibility as possible.

DarthPedro replied:
Moved to editable property with a default.

egil commented on docs/site/docs/mocking/mocking-auth.md:
```diff
@@ -1,8 +1,132 @@
-﻿---
+---
 uid: mocking-auth
```
```suggestion
uid: faking-auth
```
Please do a search/repleace in the docs for `xref:mocking-auth` and replace it with `xref:faking-auth`

egil commented on src/bunit.web/TestDoubles/Authorization/MissingFakeAuthorizationException.cs:
```diff
@@ -0,0 +1,28 @@
+			: base($"This test requires {serviceName} to be supplied, because the component under test uses authentication/authorization during the test. You can fix this by calling TestContext.Services.AddAuthorization with appropriate values. More information can be found in the documentation.")
+		{
+			this.ServiceName = serviceName;
+			this.HelpLink = "https://bunit.egilhansen.com/docs/mocking/mocking-auth.html";
```
With our decision to collect these test doubles under "TestDoubles", I think for consistency sake I will rename the mocking section in the docs to "test-doubles" as well. I.e. urls will change.

DarthPedro replied:
You already made this naming change to the documentation folder, so resolving this item.

egil commented:
FYI: I have updated the dev branch with test-doubles as the name for the documentation section.

DarthPedro commented on docs/site/docs/mocking/mocking-auth.md:
````diff
@@ -1,8 +1,132 @@
+
+You can use AuthorizeView in a component to show different content based on the user's authentication/authorization state. To set your component up for testing, you can do the following:
+
+```c#
````
This example component and tests are actually in the bunit.web.tests. They are some of the tests I used to verify the components. Would you like me to move them into a sample project or link to the test project?

egil replied:
See the markdown for this page as an example for how to include references to files in docfx markdown: https://bunit.egilhansen.com/docs/providing-input/inject-services-into-components.html

egil replied:
I have pushed an update to the DEV branch which includes an docs/README.md file that explains how to build and view the docs locally, and where things go. Hopefully that helps.

DarthPedro replied:
wrapped up documentation changes, I think.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/FakeAuthorizationPolicyProviderTest.cs:
```diff
@@ -0,0 +1,55 @@
+			// assert
+			Assert.NotNull(policy);
+			Assert.Equal(1, policy.AuthenticationSchemes.Count);
+			Assert.Equal("TestScheme:FooBar", policy.AuthenticationSchemes[0]);
```
That is just a scheme name that is in the FakeAuthorizationPolicyProvider. It follows the naming pattern for policies - SchemeName:PolicyName. But it is not settable by the caller at this point.

egil replied:
> That is just a scheme name that is in the FakeAuthorizationPolicyProvider. It follows the naming pattern for policies - SchemeName:PolicyName. But it is not settable by the caller at this point.

Can it be excluded completely, so users do not have to have a magic string in their tests, if they want to de a similar assertion like you are doing here?

Or can we a settable property with a default value that users can read/access in their tests?

DarthPedro replied:
Made the policy scheme name a settable property so that users can change it if required. Defaults to TestScheme for most usages.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/FakeAuthorizationServiceTest.cs:
```diff
@@ -0,0 +1,112 @@
+			var service = new FakeAuthorizationService();
+
+			// act
+			var result = service.NextResult;
```
NextResult is the current authorization state of FakeAuthorizationState. When Blazor calls AuthorizeAsync, this is what will be returned. It allows the user to change the authorization state when they need to. In most cases, it is configured through AddAuthorization.

egil replied:
Is the only reason to expose the property to allow users to change the state?

In that case, I prefer not to expose an implementation detail and instead have to methods that either sets it as authed or unauthed. Much easier for users I think.

DarthPedro replied:
Replaced NextResult with CurrentState property that uses the new AuthorizationState enum. That should make it easier for users to understand.

DarthPedro replied:
Also added an explicit SetAuthorizationState to FakeAuthorizationService for users to call to make it more understandable.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/SimpleAuthView.razor:
```diff
@@ -0,0 +1,16 @@
+@using Microsoft.AspNetCore.Components.Authorization
```
I will look at them both and see.

DarthPedro replied:
Yes, they are basically the same, so I am deleting the LoginDisplay component and tests.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/SimpleAuthViewTest.cs:
```diff
@@ -0,0 +1,91 @@
+
+namespace Bunit.TestDoubles.Authorization
+{
+	public class SimpleAuthViewTest
```
Yes, SimpleAuthView is the component being tested. Would AuthorizationTests be a better name? The tests are verifying how our fake services work with CascadingAuthenticationState and AuthorizeView.

egil replied:
> The tests are verifying how our fake services work with CascadingAuthenticationState and AuthorizeView.

That's what I thought. Then I think the test class should be named accordingly. I'll leave it to you to figure out a good name, naming is after all 😉

DarthPedro replied:
Renaming class to AuthorizationTest.

egil commented:
Replied to a few comments. Sorry for brevity, they are written on my mobile.

DarthPedro commented:
No worries Egil... :)

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/LoginDisplayTest.cs:
```diff
@@ -0,0 +1,58 @@
+			// assert
+			cut.MarkupMatches(@"<a href=""authentication/login"">Log in</a>");
+		}
+	}
```
Got rid of LoginDisplay component and tests since they were redundant. And SimpleAuthView already tests auth state change from unauth to auth states.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/FakeAuthorizationExtensionsTest.cs:
```diff
@@ -0,0 +1,89 @@
+				new FakePrincipal { Identity = new FakeIdentity { Name = "DarthPedro" }, Roles = roles });
+
+			// act
+			sp.AddAuthorization("DarthPedro", true, roles);
```
Replaced boolean with AuthorizationState enum.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/FakeAuthorizationExtensionsTest.cs:
```diff
@@ -0,0 +1,89 @@
+				new FakePrincipal { Identity = new FakeIdentity { Name = "DarthPedro" } });
+
+			// act
+			sp.AddAuthorization();
```
Renamed it to AddTestAuthorization.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/FakeAuthenticationStateProviderTest.cs:
```diff
@@ -0,0 +1,62 @@
+			var stateChangeHandled = false;
+			authProvider.AuthenticationStateChanged += e => stateChangeHandled = true;
+
+			var newState = FakeAuthenticationStateProvider.CreateAuthenticationState("NewUser");
```
Changed TriggerAuthenticationStateChange to take a new user or be empty to either authenticate or unauthenticate the user, so tests don't have to create the new state.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/SimpleAuthViewTest.cs:
```diff
@@ -0,0 +1,91 @@
+			var authService = (FakeAuthorizationService)ctx.Services.GetRequiredService<IAuthorizationService>();
+			authService.NextResult = AuthorizationResult.Success();
+
+			cut.Render();
```
Made the recommended changes. Simplified the way user changes state between authorized and unauthorized. It was a little more complicated than your suggestion because there are 2 services that need to change (not just AuthenticationStateProvider. So I added FakeAuthorizationExtension.UpdateTestAuthorizationState to change both in one shot.

Have a look at the changes and see if that works with your expectations for the API.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/FakeAuthorizationExtensionsTest.cs:
```diff
@@ -0,0 +1,89 @@
+			using var sp = new TestServiceProvider();
+			var roles = new List<string> { "User" };
+			var principal = new ClaimsPrincipal(
+				new FakePrincipal { Identity = new FakeIdentity { Name = "DarthPedro" }, Roles = roles });
```
Created a factory method to encapsulate the ClaimsPrincipal creation.

egil replied:
Is this low level stuff / internal bits, that most users never have to deal with? In that case, it can stay like this.

egil commented on src/bunit.web/TestDoubles/Authorization/MissingFakeAuthorizationException.cs:
```diff
@@ -0,0 +1,28 @@
+			: base($"This test requires {serviceName} to be supplied, because the component under test uses authentication/authorization during the test. You can fix this by calling TestContext.Services.AddAuthorization with appropriate values. More information can be found in the documentation.")
+		{
+			ServiceName = serviceName;
+			HelpLink = "https://bunit.egilhansen.com/docs/mocking/mocking-auth.html";
```
```suggestion
			HelpLink = "https://bunit.egilhansen.com/docs/test-doubles/faking-auth.html";
```

egil commented on src/bunit.web.tests/TestDoubles/Authorization/AuthorizationTest.cs:
```diff
@@ -0,0 +1,108 @@
+		{
+			// arrange
+			using var ctx = new TestContext();
+			ctx.Services.AddTestAuthorization();
+			var authProvider = ctx.Services.GetRequiredService<AuthenticationStateProvider>();
+
+			// start off unauthenticated.
+			var cut = ctx.RenderComponent<SimpleAuthView>();
+			cut.MarkupMatches("Not authorized?");
+
+			// act
+			ctx.Services.UpdateTestAuthorizationState("TestUser004", AuthorizationState.Authorized);
+
+			cut.Render();
```
This doesn't look right to me. 

`Services.AddTestAuthorization(...)` should return a special object that allows the user to interact with the fake authentication and authorization logic you are building, i.e. like [`AddMockJsRuntime`](https://github.com/egil/bUnit/blob/main/src/bunit.web/Mocking/JSInterop/MockJsRuntimeExtensions.cs/#L16). 

That object, e.g. `FakeAuthStateManager` (probably not the best name), can then have methods that like `UpdateState(...)`, which changes the state. It can also have properties that shows the current state and provides access to the magic strings used within, e.g. `TestSchema`. 

Note, that its `UpdateState` doesn't have to be `UpdateTestAuthorizationState` because the method exists on an type named `FakeAuthStateManager` or similar, which gives the method name context.

That said, methods like `UpdateTestAuthorizationState` (and `UpdateState` on a `FakeAuthStateManager`) is also bugging me. I think it is because the method implicitly does two things, either set an authenticated or unauthenticated state, depending on whether there are parameters present or not. The result is, that when you read a test, and see a call to `UpdateTestAuthorizationState()` without any parameters, you have to know that it means "not authenticated". Thus, instead of `UpdateTestAuthorizationState`/`UpdateState` I would much rather have properties or methods with good names that makes their effect obvious to the reader.

The `FakeAuthStateManager` can also provide access to the fake principle, roles, etc.. 

Makes sense?

```C#
var authManager = ctx.Services.AddTestAuthorization();
authManager.IsAuthenticated.ShouldBeFalse(); // only here for demonstration purposes
authManager.Authorization.ShouldBe(AuthorizationState.Unauthorized); // only here for demonstration purposes
                        
// start off unauthenticated.
var cut = ctx.RenderComponent<SimpleAuthView>();
cut.MarkupMatches("Not authorized?");

// act
authManager.Authenticate("TestUser004", AuthorizationState.Authorized); // is it a good name?

...
```

Another thing. `Services` should not have an UpdateTestAuthorizationState method, it does not logically belong to a services collection. Services collections allows you to add services (register) and get them, nothing else.

Sorry for going back and forth on this. I guess we should have had a API design discussion before we got too deep into the coding.

DarthPedro replied:
I misunderstood the pattern you use for AddMockJSRuntime, and what it returned. I'm used to the Services.AddXxx extensions pattern in ASP.NET Core, which usually returns the ServiceProvider, so calls can be chained together. I don't have a strong conviction either way, but I would not expect an object returned from Service.Add methods.

I will try some of the changes to AddTestAuthorization that you suggested above and see how it looks.

I also felt weird adding UpdateTestAuthorizationState to the Services collection, but didn't have another good place to put the update methods.

DarthPedro replied:
Ok, create an AuthorizationAdaptor class returned by the extension method that wraps the auth services and some state. Simplified the extension methods. AuthorizationAdaptor class has an Authenticate and Unauthenticate method. This cleaned up some of the tests and the extension methods. Let me know what you think.

I changed the FakeAuthorizationExtensions.AddTestAuthorization to be a single method that registers the auth services with the service provider and sets up in the unauthenticated state by default. Then, the user calls AuthorizationAdaptor.Authenticate if a user is required. What do you think about that pattern?

egil replied:
I just read through the new tests and it looks much cleaner. I am not 100% about the naming of the type and the methods. What do you think about `TestAuthorizationContext`? `

`Authenticate` and `Unauthenticate` shoulds like something we tell something to do, i.e. commands, but that is not really the case here, here it is more about setting an auth state. Any suggestions for better names.

I agree that normally `Add*` methods on service collections should not return something, but in this case I think we can deviate from the CQS principle, as it makes tests simpler. I expect to use that pattern for all test doubles that needs to be registered through the service collection, so users should become familiar with the pattern.

DarthPedro replied:
Yes, I like TestAuthorizationContext for the class name.
For the method names, I had ChangeAuthorizationState before with overloads, but you didn't like having two separate methods with the same name. I could have just one ChangeAuthorizationState where you call ChangeAuthorizationState(null, AuthorizationState.Unauthorized), but passing null or empty string as the user name to mean your unauthenticated won't be easy to understand for users either.

How about something like ChangeStateAuthenticated and ChangeStateUnauthenticated methods? The names better reflect what the methods are doing. And it's clear to the user what each method does. Thoughts?

egil replied:
> How about something like ChangeStateAuthenticated and ChangeStateUnauthenticated methods? The names better reflect what the methods are doing. And it's clear to the user what each method does. Thoughts?

The `ChangeState` part of the names does not really add much I think. What about simply calling the methods `SetAuthenticated` and `SetUnauthenticated`?

DarthPedro replied:
Ok, I will make those method name changes.

DarthPedro replied:
rename complete

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/LoginDisplay.razor:
```diff
@@ -0,0 +1,22 @@
+@using Microsoft.AspNetCore.Components.Authorization
```
Moved SimpleAuthView component to bunit.testassets/SampleComponents... got rid of LoginDisplay component.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/FakeAuthorizationServiceTest.cs:
```diff
@@ -0,0 +1,112 @@
+			Assert.False(result.Succeeded);
+		}
+
+		[Fact(DisplayName = "Get exception from invoking placeholder AuthorizationService methods.")]
+		public async Task Test005()
+		{
+			// arrange
+			using var ctx = new TestContext();
+			var authService = ctx.Services.GetRequiredService<IAuthorizationService>();
+			IEnumerable<IAuthorizationRequirement>? requirements = null;
+
+			// act
+			var ex = await Assert.ThrowsAsync<MissingFakeAuthorizationException>(() =>
+				authService.AuthorizeAsync(null, string.Empty, requirements));
+
+			// assert
+			Assert.Equal("IAuthorizationService", ex.ServiceName);
+		}
+
+		[Fact(DisplayName = "Get exception from invoking placeholder AuthorizationService methods.")]
+		public async Task Test006()
+		{
+			// arrange
+			using var ctx = new TestContext();
+			var authService = ctx.Services.GetRequiredService<IAuthorizationService>();
+
+			// act
+			var ex = await Assert.ThrowsAsync<MissingFakeAuthorizationException>(() =>
+				authService.AuthorizeAsync(null, string.Empty, "testPolicy"));
+
+			// assert
+			Assert.Equal("IAuthorizationService", ex.ServiceName);
+		}
```
There wasn't an easy way to turn these two into a Theory. AuthorizeAsync is overloaded and the type of the Requirement or policy string is important to which method gets called. But I didn't see a way to use a Theory to pass a typed parameter to the test method and it cannot be object for the overload to work.

Thought about using generics, but looks like xUnit doesn't support generics for Theory data... https://github.com/xunit/xunit/issues/1378.

DarthPedro replied:
So keeping as two separate Facts for now.

egil commented on src/bunit.web.tests/TestDoubles/Authorization/AuthorizationTest.cs:
```diff
@@ -0,0 +1,107 @@
+			// act
+			authAdaptor.Authenticate("TestUser004", AuthorizationState.Authorized);
+
+			cut.Render();
```
Is there a way to get rid of this? Should the `<CascadingAuthenticationState>` not notice the change and automatically cause a re-render in its child components?

DarthPedro replied:
It does need the explicit render call on the auth state change. I tried without that and a couple of different options, but the render appears to be needed.

DarthPedro commented on docs/site/docs/test-doubles/faking-auth.md:
````diff
@@ -1,8 +1,132 @@
+
+You can use AuthorizeView in a component to show different content based on the user's authentication/authorization state. To set your component up for testing, you can do the following:
+
+```c#
+@using Microsoft.AspNetCore.Components.Authorization
+
+<CascadingAuthenticationState>
+	<AuthorizeView>
+		<Authorized>
+			Authorized!
+		</Authorized>
+		<NotAuthorized>
+			Not authorized?
+		</NotAuthorized>
+	</AuthorizeView>
+</CascadingAuthenticationState>
+```
````
I already have that SimpleAuthView.razor file and the corresponding tests in the bunit.testassets project (so that they run as part of unit tests). Do we want to duplicate them in the docs/samples folder as well? Or can we link to the existing files in the src test folder?

egil replied:
Yes, I would like to have docs specific samples in its own place. They should be able to evolve independently of each other.

DarthPedro replied:
Ok, I will make all of the doc/sample changes in one pass after we close down on the code changes.

DarthPedro replied:
done

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/AuthorizationAdaptor.cs:
```diff
@@ -0,0 +1,73 @@
+	/// Root authorization service that manages different authentication/authorization state
+	/// in the system.
+	/// </summary>
+	public class AuthorizationAdaptor
```
Renamed to TestAuthorizationContext.

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/AuthorizationAdaptor.cs:
```diff
@@ -0,0 +1,73 @@
+		/// <param name="userName">User name for the principal identity.</param>
+		/// <param name="state">Authorization state.</param>
+		/// <param name="roles">Roles for the claims principal.</param>
+		public void Authenticate(string userName, AuthorizationState state, IEnumerable<string>? roles = null)
```
I will give it an AuthorizationState.Authorized default.

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/AuthorizationAdaptor.cs:
```diff
@@ -0,0 +1,73 @@
+		/// <summary>
+		/// Gets the set of roles for the current user.
+		/// </summary>
+		public IEnumerable<string>? Roles { get; private set; } = null;
```
Unfortunately you can't use Array.Empty<string>() as the default for the roles parameter in the Authenticate method, so that has to be nullable. I could have the property be non-nullable, but the method with default roles would still need to be nullable. Is that ok?

DarthPedro replied:
Made property non-nullable, but Authenticate method remains IEnumerable<string>? so that we can pass a default value. Method converts null to Array.Empty<string>() when that happens.

egil replied:
Yep, that's the right approach.

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/AuthorizationAdaptor.cs:
```diff
@@ -0,0 +1,73 @@
+		public void Unauthenticate()
+		{
+			IsAuthenticated = false;
+			Roles = null;
```
made this change.

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/FakePrincipal.cs:
```diff
@@ -0,0 +1,51 @@
+		/// <returns>Returns that this principal is not in any role.</returns>
+		public bool IsInRole(string role)
+		{
+			if (Roles == null)
```
made Roles non-nullable so removed this check altogether.

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/FakeAuthorizationService.cs:
```diff
@@ -0,0 +1,75 @@
+			AuthorizeCalls.Add((user, resource, requirements));
+
+			// The FakeAuthorizationService doesn't apply any authorization requirements.
+			// It just returns whatever the user specified in the NextResult property.
```
removed comment references to NextResult

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/FakePrincipal.cs:
```diff
@@ -0,0 +1,51 @@
+		/// <summary>
+		/// Gets or sets the set of roles this user is authorized for.
+		/// </summary>
+		public IEnumerable<string>? Roles { get; set; }
```
made Roles property non-nullable

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/FakePrincipal.cs:
```diff
@@ -0,0 +1,51 @@
+		public static ClaimsPrincipal CreatePrincipal(string userName, IList<string>? roles = null)
+		{
+			var principal = new ClaimsPrincipal(
+				new FakePrincipal { Identity = new FakeIdentity { Name = userName }, Roles = roles });
```
done

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/FakeAuthenticationStateProvider.cs:
```diff
@@ -0,0 +1,91 @@
+		public static Task<AuthenticationState> CreateAuthenticationState(string username, IEnumerable<string>? roles = null)
+		{
+			var identity = new FakeIdentity { Name = username };
+			var testPrincipal = new FakePrincipal { Identity = identity, Roles = roles };
```
done

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/FakeAuthenticationStateProvider.cs:
```diff
@@ -0,0 +1,91 @@
+		/// </summary>
+		/// <param name="userName">Identity's user name.</param>
+		/// <param name="roles">Roles that this user principal has.</param>
+		public FakeAuthenticationStateProvider(string userName, IList<string>? roles = null)
```
done

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/AuthorizationAdaptor.cs:
```diff
@@ -0,0 +1,73 @@
+	{
+		private readonly FakeAuthorizationService authService = new FakeAuthorizationService();
+		private readonly FakeAuthorizationPolicyProvider policyProvider = new FakeAuthorizationPolicyProvider();
+		private readonly FakeAuthenticationStateProvider authProvider = new FakeAuthenticationStateProvider();
```
done

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/AuthorizationAdaptor.cs:
```diff
@@ -0,0 +1,73 @@
+		/// <summary>
+		/// Gets whether user is authenticated.
+		/// </summary>
+		public bool IsAuthenticated { get; private set;  } = false;
```
fixed

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/FakeAuthorizationPolicyProvider.cs:
```diff
@@ -0,0 +1,56 @@
+		/// <summary>
+		/// Gets or sets the policy scheme name (if user wants to set it). Defaults to TestScheme.
+		/// </summary>
+		public string PolicySchemeName { get; set; } = "TestScheme";
```
added check

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/FakeIdentity.cs:
```diff
@@ -0,0 +1,25 @@
+		/// <summary>
+		/// Gets the test authentication type.
+		/// </summary>
+		public string AuthenticationType { get; set; } = "Test";
```
added check

egil commented on src/bunit.web/TestDoubles/Authorization/FakeAuthorizationService.cs:
```diff
@@ -0,0 +1,75 @@
+		/// <summary>
+		/// Gets a list of the AuthorizeAsync calls that were made to this service.
+		/// </summary>
+		public List<(ClaimsPrincipal user, object resource, IEnumerable<IAuthorizationRequirement> requirements)> AuthorizeCalls { get; }
```
How would users get to this? Should it be part of testauthcontect?

DarthPedro replied:
Exposed it as passthrough property in TestAuthorizationContext.

egil commented:
@DarthPedro did a few quick refactorings and checked out your work locally on my machine.

I think you are right, we cannot get out of calling Render() in tests after changing state, The Blazor team does that in its own tests: https://source.dot.net/#Microsoft.AspNetCore.Components.Authorization.Tests/CascadingAuthenticationStateTest.cs,69

The only thing in the code I am still a bit unsure about is the need for `AuthorizeCalls`. What are you reasoning for wanting to do that kind of assertion?

DarthPedro commented:
@egil I pulled down your changes and in general they look good. But I am concerned about your change for string property setters. It used to be:
	set => _policySchemeName = value.VerifyRequiredValue();
But you changed it to:
	set => _policySchemeName = value ?? throw new ArgumentNullException(nameof(value));

However, VerifyRequiredValue did a string.IsNullOrEmpty check not just a null check so this changed the behavior and we don't want users setting those properties to empty strings. Should I change it back? If you did not like the extension method, I can put the check in the setter without the extension method.

DarthPedro commented:
For `AuthorizeCalls`, it was helpful in debugging how AuthorizeView and CascadingAuthenticationState worked and calls that were made to IAuthorizationService. But perhaps it was more helpful in the investigation phase than in user tests.

I can remove that if you don't feel it will help bUnit users.

egil commented:
> However, VerifyRequiredValue did a string.IsNullOrEmpty check not just a null check so this changed the behavior and we don't want users setting those properties to empty strings. Should I change it back? If you did not like the extension method, I can put the check in the setter without the extension method.

@DarthPedro I noticed that since I broke your FakeIdentity tests, so I changed the check in that class and the tests should be passing again. If there are other places where it is not a null check but also an empty string check, then please fix my mistake. I do prefer it without the extension method, since it is so few places we need it in, right now at least, and that would be another API that would need careful naming and design :)

> For `AuthorizeCalls`, it was helpful in debugging how AuthorizeView and CascadingAuthenticationState worked and calls that were made to IAuthorizationService. But perhaps it was more helpful in the investigation phase than in user tests.
> 
> I can remove that if you don't feel it will help bUnit users.

I just want to make sure we do not include something that is not needed, since it will have to be supported in the future. So if we cannot imagine a good case for it, I would remove it for now. If somebody raises and issue and has a good use case, I will be happy to include it.

DarthPedro commented:
Ok, latest change added back empty string check on property setter. And removed the AuthorizeCalls property... we can always add it back if we find a good use case.

I think I've address all of your code PR comments. Going to wrap up the documentation changes next.

DarthPedro commented:
Question on the sample app: I've added SimpleAuthView to the SampleApp project and trying add a corresponding test class to the SampleApp.Tests project. However, the test project references the bUnit nuget beta package. There isn't a nuget build with my changes yet that I can reference in the SampleApp.Tests project, so how can I add the sample tests for my work? (chicken-egg problem)

egil commented:
You do not need to update the `/sample/` sample test app. Just add the "doc samples" in `/docs/samples/...`.

DarthPedro commented:
Haha, I was just typing... nevermind, I was in the wrong sample app. :)
We should probably remove the /sample folder at some point.

egil commented:
Yeah, I have been considering removing the sample app completely, now that the docs are finally getting some love.

DarthPedro commented on docs/site/docs/test-doubles/faking-auth.md:
````diff
@@ -1,8 +1,132 @@
+
+The following code tests the component with an unauthenticated user. To setup an authenticated user, just call AddTestAuthorization with no parameters.
+
+```c#
+[Fact(DisplayName = "AuthorizeView with unauthenticated user")]
+public void Test001()
+{
+	// arrange
+	using var ctx = new TestContext();
+	ctx.Services.AddTestAuthorization();
+
+	// act
+	var cut = ctx.RenderComponent<SimpleAuthView>();
+
+	// assert
+	cut.MarkupMatches("Not authorized?");
+}
+```
````
done

DarthPedro commented on docs/site/docs/test-doubles/faking-auth.md:
````diff
@@ -1,8 +1,132 @@
+
+Now we can test with an authenticated and authorized user by calling AddTestAuthorization with a user name and authorization flag.
+
+```c#
+[Fact(DisplayName = "AuthorizeView with authenticated and authorized user")]
+public void Test002()
+{
+	// arrange
+	using var ctx = new TestContext();
+	ctx.Services.AddTestAuthorization("TestUser", AuthorizationState.Authorized);
+
+	// act
+	var cut = ctx.RenderComponent<SimpleAuthView>();
+
+	// assert
+	cut.MarkupMatches("Authorized!");
+}
+```
````
done

DarthPedro commented on CHANGELOG.md:
````diff
@@ -9,6 +9,43 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
+</CascadingAuthenticationState>
+```
+Then define your test method to specify the authorization state with a user name, and run your test.
+```c#
+		[Fact(DisplayName = "AuthorizeView with authenticated and authorized user")]
+		public void Test002()
+		{
+			// arrange
+			using var ctx = new TestContext();
+			ctx.Services.AddTestAuthorization("TestUser", AuthorizationState.Authorized);
+
+			// act
+			var cut = ctx.RenderComponent<SimpleAuthView>();
+
+			// assert
+			cut.MarkupMatches("Authorized!");
+		}
````
done

DarthPedro commented:
I think that may be all of the changes. :)

egil commented:
Thanks. I'll take a look later and tweak the docs to match the style needed. If nothing is unclear I'll merge it in later. 

Many thanks for the effort on this.

DarthPedro commented:
Ok, you're welcome. Glad to help.

egil commented:
@DarthPedro while tweaking the docs, I was changing around some of the tests, and I think I have a scenario that fails but should not.

With this component:

```html
@using Microsoft.AspNetCore.Components.Authorization
@inject AuthenticationStateProvider AuthenticationStateProvider

@if (isAuthenticated)
{
  <p>User name: @userName</p>
  <CascadingAuthenticationState>
	  <AuthorizeView>
		  <Authorized>
			  <p>State: Authorized</p>
		  </Authorized>
		  <NotAuthorized>
			  <p>State: Not authorized</p>
		  </NotAuthorized>
	  </AuthorizeView>
  </CascadingAuthenticationState>
}

@code
{
  bool isAuthenticated = false;
  string userName;

  protected override async Task OnInitializedAsync()
  {
    var state = await AuthenticationStateProvider.GetAuthenticationStateAsync();
    isAuthenticated = state.User.Identity.IsAuthenticated;
    userName = state.User.Identity.Name;
  }
}
```

This test fails:

```c#
[Fact(DisplayName = "SimpleAuthView switch from unauthorized to authorized.")]
public void Test004()
{
  // arrange
  using var ctx = new TestContext();
  
  var authContext = ctx.Services.AddTestAuthorization();

  // start off unauthenticated.
  var cut = ctx.RenderComponent<SimpleAuthView>();
  cut.MarkupMatches("");

  // act - set authorized state
  authContext.SetAuthorized("TestUser", AuthorizationState.Authorized);
  cut.Render();

  // assert
  cut.WaitForAssertion(() => 
    cut.MarkupMatches(@"<p>User name: TestUser</p>
                      <p>State: Authorized</p>")
  );
}
```

This should work, right?

egil commented:
Ahh nevermind, I changed the component so it uses `OnParametersSetAsync` instead of `OnInitializedAsync`, since the first is rerun after each render.

Anyway, I am moving this weekend with my family, so cannot promise I can finish the review before monday. Hope that is OK.

egil commented:
btw. the assertion seems to work fine without `cut.WaitForAssertion` on my machine.

egil commented:
Do we provide a way to set a policy? E.g. to support this:

```
<AuthorizeView Policy="content-editor">
    <p>You can only see this if you satisfy the "content-editor" policy.</p>
</AuthorizeView>
```

DarthPedro commented:
Currently that works, the FakePolicyProvider always passes the authorization policy check and returns the name that was requested (I just tried it in a test). It's a simple pass-thru that always succeeds.

However, the negative test scenario doesn't work, if you want to test what is rendered when the user does not have the required policy. I can look at adding a Policies list, similar to Roles, that only passes the FakePolicyProvider if the requested is part of the Policies list.

Do you want that as part of this PR or as an enhancement after?

egil commented:
I think it should be party of this release. Another though has crossed my mind... The roles list could be a `params string[]` in the SetAuthorized method, making it easier to pass just one role, which I think is a quite likely scenario.

DarthPedro commented:
Ok, changed SetAuthorized to take params string[] roles.
Added TestAuthorizationContext.SetAuthorizationPolicy to allow users to add policies when needed to test their components (defaults to no policies). Also made that method take param string[] policies, since usually you're only testing one.

Add SimpleAuthViewWithPolicy component and tests to samples project so that it can be documented.

egil commented on src/bunit.web/TestDoubles/Authorization/TestAuthorizationContext.cs:
```diff
@@ -0,0 +1,114 @@
+		/// <param name="userName">User name for principal identity.</param>
+		/// <param name="roles">Roles for the user.</param>
+		/// <returns>ClaimsPrincipal created from this data.</returns>
+		public static ClaimsPrincipal CreatePrincipal(string userName, IEnumerable<string>? roles = null)
```
Is there a need for this to be public?

egil commented on src/bunit.web/TestDoubles/Authorization/TestAuthorizationContext.cs:
```diff
@@ -0,0 +1,114 @@
+		/// Sets the auhtorization policies supported for the current user.
+		/// </summary>
+		/// <param name="policies">Supported authorization policies.</param>
+		public void SetAuthorizationPolicy(params string[] policies)
```
It seems like this turns policies into 2nd class citizens compared to roles.

Perhaps there should be a `SetRoles(params string[] roles)` and `SetPolicies(params string[] policies)` instead, and not make roles settable through `SetAuthorized(...)`?

If we make it builder pattern like, e.g. by having set methods return `TestAuthorizationContext`, then it can be done in a one-liner in tests:

```c#
authCtx.SetAuthorized("egil").SetRoles("reviewer");
// or...
authCtx.SetAuthorized("egil").SetPolicies("reviewer");
```

I assume that in most cases, one would never use both at the same time, although that is possible.

DarthPedro replied:
Added SetRoles method, and made the TestAuthorizationContext methods chainable. Updated sample code to use new api.

egil commented on src/bunit.web.tests/TestDoubles/Authorization/AuthorizationTest.cs:
```diff
@@ -0,0 +1,140 @@
+
+namespace Bunit.TestDoubles.Authorization
+{
+	public class AuthorizationTest
```
Do we need a test where `CUT` has a 

```
    [CascadingParameter]
    private Task<AuthenticationState> authenticationStateTask { get; set; }
```
?

DarthPedro replied:
I don't think we need that separate test. That's what <CascadingAuthenticationState> already does. Plus you still need to call TestAuthorizationContext.SetAuthorized to set up the other required services to make AuthorizeView tests work.

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/AuthorizationContextTest.cs:
```diff
@@ -0,0 +1,77 @@
+			// arrange
+			using var sp = new TestServiceProvider();
+			var authContext = sp.AddTestAuthorization();
+			authContext.SetAuthorized("DarthPedro", AuthorizationState.Authorized, "User");
```
done

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/AuthorizationContextTest.cs:
```diff
@@ -0,0 +1,77 @@
+			// assert
+			Assert.True(authContext.IsAuthenticated);
+			Assert.Equal(AuthorizationState.Authorized, authContext.Authorization);
+			Assert.Equal(new List<string> { "User" }, authContext.Roles);
```
done

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/FakeAuthorizationPolicyProviderTest.cs:
```diff
@@ -0,0 +1,86 @@
+			var provider = new FakeAuthorizationPolicyProvider();
+
+			// act
+			var policy = await provider.GetFallbackPolicyAsync().ConfigureAwait(false);
```
removed

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/FakeAuthorizationPolicyProviderTest.cs:
```diff
@@ -0,0 +1,86 @@
+			provider.SetPolicies("TestScheme", new List<string> { "FooBar" });
+
+			// act
+			var policy = await provider.GetPolicyAsync("FooBar").ConfigureAwait(false);
```
removed

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/FakeAuthorizationPolicyProviderTest.cs:
```diff
@@ -0,0 +1,86 @@
+			provider.SetPolicies("TestScheme", new List<string> { "FooBar" });
+
+			// act
+			var policy = await provider.GetPolicyAsync("OtherPolicy").ConfigureAwait(false);
```
removed

DarthPedro replied:
removed

egil commented:
I pushed some changes to the PR including refactored docs to fit the general style.

While documenting policies and roles i stumples onto something I think is a bug. I added a bunch of extra tests to the project, but in essens, the problem is content limited by roles and policies are always shown, even when inside a `<AuthorizeView Roles="superuser">` component.

DarthPedro commented:
I don't see any new pushed commits from you to the PR. Did you forget to push them? ;)

egil commented:
Weird, I did not. But maybe I don't have the right `remote` set. Let me check and get back to you.

egil commented:
Ahh, I managed to push a new branch to this repo with the changes: https://www.github.com/egil/bUnit/tree/dev-authorization-fakes/

Not sure how I add these to your PR without pushing directly to your local repo. Suggestions?

DarthPedro commented:
That's weird. You pushed a commit into this PR earlier: https://github.com/egil/bUnit/pull/151/commits/3e4b04d70d82ac6eee1965c8264d28ce772bd938. Did you do that in my forked repo?

egil commented:
Hmm nop. But I did pull the PR down using the `gh pr pull` command... Maybe that set something up. I'll have to check tonight after the kids are down for the night.

egil commented:
OK, I managed to push it to this PR instead. It seems that when you use `gh pr checkout 151` to review locally, you cannot change to another branch and come back again for a later review.

DarthPedro commented:
Ok, I have your changes now. I will look at the failing tests.

DarthPedro commented:
Fixed the AuthorizationService.AuthorizeAsync to validate authorization state based on requirements and either policies, roles, or authorization state. This covers all of the states we have, I believe.

I'll look at a SetClaims method as well though I'm not sure we'll be able to fake of the behavior that ClaimsPrincipal.Claims does.

egil commented on src/bunit.web/TestDoubles/Authorization/FakeAuthorizationService.cs:
```diff
@@ -13,11 +16,13 @@ namespace Bunit.TestDoubles.Authorization
 	{
 		private AuthorizationState _currentState = AuthorizationState.Authorized;
+		private IEnumerable<string>? _supportedPolicies;
+		private IEnumerable<string> _supportedRoles = new List<string>();
```
```suggestion
		private IEnumerable<string> _supportedRoles = Array.Empty<string>();
```

DarthPedro replied:
updated

egil commented on src/bunit.web/TestDoubles/Authorization/FakeAuthorizationService.cs:
```diff
@@ -32,6 +37,24 @@ public void SetAuthorizationState(AuthorizationState state)
+		/// <param name="roles">List of supported roles.</param>
+		public void SetRoles(IEnumerable<string> roles)
+		{
+			_supportedRoles = roles;
```
```suggestion
			_supportedRoles = roles ?? Array.Empty<string>();
```

DarthPedro replied:
changed

egil commented on src/bunit.web/TestDoubles/Authorization/TestAuthorizationContext.cs:
```diff
@@ -85,6 +86,8 @@ public TestAuthorizationContext SetRoles(params string[] roles)
 			_authProvider.TriggerAuthenticationStateChanged(UserName, Roles);
 
+			_authService.SetRoles(Roles);
+
```
```suggestion
			_authService.SetRoles(Roles);
			_authProvider.TriggerAuthenticationStateChanged(UserName, Roles);
```
Should these two lines not be in this order? I.e. first set roles in auth service and then notify?

DarthPedro replied:
These are setting two separate, independent services, so order isn't important as long as both happen before method returns. But swapped calls if that makes it easier to understand.

DarthPedro commented:
I got both SetClaims and the AuthorizeView.Authorizing state working. Authorizing was a little tricky because how to get into that state is not documented. But I figured it out looking through the AuthorizeViewCore code.

egil commented on src/bunit.web/TestDoubles/Authorization/FakeAuthorizationService.cs:
```diff
@@ -0,0 +1,145 @@
+			{
+				result = VerifyRequiredRoles(requirements);
+			}
+			else if (_supportedPolicies != null)
```
```suggestion
			else if (_supportedPolicies != null)
```
```suggestion
			else if (_supportedPolicies is {})
```

DarthPedro replied:
done

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/AuthorizationTest.cs:
```diff
@@ -210,5 +211,26 @@ public void Test012()
+			var cut = ctx.RenderComponent<SimpleAuthViewWithClaims>();
+
+			// assert
+			cut.MarkupMatches(@"<div>Authorized!</div>
+								<div>Name: TestUser</div>
+								<div>Email: user@test.com</div>
+								<div>Id: 5d5fa9c1-abf9-4ed6-8fb0-3365382b629c</div>");
```
done

DarthPedro commented on src/bunit.web/TestDoubles/Authorization/FakeAuthenticationStateProvider.cs:
```diff
@@ -80,13 +82,22 @@ public void TriggerAuthenticationStateChanged()
 			var testPrincipal = new FakePrincipal { Identity = identity, Roles = roles ?? Array.Empty<string>() };
 			var principal = new ClaimsPrincipal(testPrincipal);
 
+			if (claims != null && claims.Any())
```
done

DarthPedro commented on src/bunit.web.tests/TestDoubles/Authorization/AuthorizationTest.cs:
```diff
@@ -210,5 +211,26 @@ public void Test012()
 			cut.MarkupMatches("Authorizing...");
 		}
+
+		[Fact(DisplayName = "AuthorizeView with claims")]
```
Added negative test.

egil commented:
@DarthPedro I changed the code a bit to make it possible to call `SetAuthorizing()` before RenderComponent. It was necessary as the code otherwise through a null pointer exception, since you were passing a `null!` in as the authentication state.

Please give my changes a review. If it looks alright, I'd say we are done!

DarthPedro commented:
Ok, that looks good. Thanks for all of the help.

egil commented:
> Ok, that looks good. Thanks for all of the help.

No thank you. You almost all the work here!
